packages/react-components/react-utilities/src/hooks/useMotionPresence.ts

```diff
@@ -0,0 +1,345 @@
+  const window = node.ownerDocument?.defaultView;
+
+  return window!.getComputedStyle(node, null);
+};
+
+/**
+ * Converts a CSS duration string to milliseconds.
+ *
+ * @param duration - CSS duration string
+ * @returns Duration in milliseconds
+ */
+function toMs(duration: string): number {
+  const trimmed = duration.trim();
+
+  if (trimmed.includes('auto')) {
+    return 0;
+  }
+
+  if (trimmed.includes('ms')) {
+    return parseFloat(trimmed);
+  }
+
+  return Number(trimmed.slice(0, -1).replace(',', '.')) * 1000;
+}
+
+/**
+ * Checks if the browser supports CSSOM.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ *
+ * @param node - DOM node
+ * @returns Whether the browser supports CSSOM
+ */
+const hasCSSOMSupport = (node: HTMLElementWithStyledMap) => {
+  /**
+   * As we are using the experimental CSSOM API, we need to check if the browser supports it.
+   * The typecast here is to allow the use of the `number` function that is not yet part of the CSSOM typings.
+   * @see https://www.npmjs.com/package/@types/w3c-css-typed-object-model-level-1
+   */
+  return Boolean(typeof CSS !== 'undefined' && (CSS as unknown as CSSWithNumber).number && node.computedStyleMap);
+};
+
+/**
+ *
+ * Gets the computed style of a given element.
+ * If the browser supports CSSOM, it will return a ComputedStyleMap object.
+ * Otherwise, it will return a CSSStyleDeclaration object.
+ */
+const getCSSStyle = (node: HTMLElementWithStyledMap): CSSStyleDeclaration | StylePropertyMapReadOnly => {
+  if (hasCSSOMSupport(node)) {
+    return node.computedStyleMap();
+  }
+
+  return getElementComputedStyle(node);
+};
+
+/**
+ * Gets the computed map property for a given element using the CSSOM API.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ *
+ * @param computedStyle - Computed style of the element
+ * @param prop - CSS property
+ * @returns Computed map property
+ */
+const getComputedMapProp = (computedStyle: StylePropertyMapReadOnly, prop: string): string[] => {
+  const props = computedStyle.getAll(prop);
+
+  if (props.length > 0) {
+    return props.map(({ value, unit }) => `${value}${unit}`);
+  }
+
+  return ['0'];
+};
+
+/**
+ * Gets the computed style property for a given element using the getComputedStyle API.
+ *
+ * @param computedStyle - Computed style of the element
+ * @param prop - CSS property
+ * @returns Computed style property
+ */
+const getComputedStyleProp = (computedStyle: CSSStyleDeclaration, prop: string): string[] => {
+  const propValue = computedStyle.getPropertyValue(prop);
+
+  return propValue ? propValue.split(',') : ['0'];
+};
+
+/**
+ * Gets the maximum duration from a list of CSS durations.
+ *
+ * @param durations - List of CSS durations
+ * @param delays - List of CSS delays
+ * @returns Maximum duration
+ */
+const getMaxCSSDuration = (durations: string[], delays: string[]): number => {
```

Comment:
You should consider reworking this to be a single loop. I wrote an example implementation in this benchmark:

https://jsbench.me/2alksrrlr2/3